docs/blog/2021-12-01-csp.md

```diff
@@ -0,0 +1,153 @@
+Google. It's pretty common to include a few additional domains (like Google) in here that you need to load things from,
+such as analytics or some third-party auth scripts.  
+
+That CSP also means that inline scripts like `<script> alert('Hi I'm a script') <script>` won't be executed, which is great for security.
+
+So, if we trust our server and Google's server not to do something malicious, we would expect to be safe.  
+
+Unfortunately, this CSP is still totally insecure, because:
+
+## CSP is swiss cheese
+Studies have found that the vast majority of sites using CSPs are still vulnerable to script injection.  Here are some of the most common attacks.
+### JSONP Bypass
+JSONP is a hacky way to load cross domain javascript in the browser from the mid 2000's, that has since been replaced by CORS.  More of a pattern than a protocol, 
+the requester is able to tell the server to wrap whatever javascript it sends back in whatever snippet you pass as a URL parameter.
+
+So this script tag:
+```html
+<script src="https://accounts.google.com/o/oauth2/revoke?callback=alert('This is a malicious alert')"></script>
+```
+Will return this:
+```javascript
+// API callback
+alert('This is a malicious alert')({
+  // ...googles intended payload
+}
+);
+```
+
+Go ahead and try [the url from that script](<https://accounts.google.com/o/oauth2/revoke?callback=alert('This is a malicious alert')>) in your browser.
+You'll quickly see the problem.  Now an attacker that can embed HTML in your page has
+the ability to execute whatever javascript they want. Any un-sanitized field that leaks HTML into your page is now an immediate vulnerability.
+
+There a great number of these vulnerable legacy endpoints, not just on googles domains but across the web.  
+Here is a list of [just some of them](https://github.com/zigoo0/JSONBee/blob/master/jsonp.txt).
+
+Because this isn't a specific protocol but just a serverside (anti)pattern, there's no way for the browser
+to detect this behavior and disable it.  The only solution is to not allow scripts to load from any vulnerable domain, and 
+to host them yourself instead.
+
+ **The bigger your website, the harder this is to deal with.**
+
+If you have a small site and immediate control over the codebase, this could be an easy problem to patch.
+
+For large sites at big companies that load analytics payloads, third party auth, and other scripts from these vulnerable domains, it can be 
+difficult to mitigate.  Heck, a lot of big companies have legacy JSONP endpoints on the _same domain as the site._  There's not much you can do about that.
+
+### File Upload Bypass
+Let's say your CSP is locked to only allow scripts from your domain.  
+
+If your site allows file uploads and downloads, including files that end in `.js`, a user can simply upload a file `malicious-code.js`,
+and load it as a script tag, directly from your domain. 
+
+```html
+<script src=www.yourdomain.com/myfiles/malicious-code.js></script>
```

Comment:
there is a caveat to this, the content type of the returned file needs to be `application/javascript` for this to work.